File: templates/Enterprise-Template/src/typescript/src/dialogs/escalate/escalateDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { DialogTurnResult, WaterfallDialog, WaterfallStepContext } from "botbuilder-dialogs";
import { BotServices } from "../../botServices";
import { EnterpriseDialog } from "../shared/enterpriseDialog";
import { EscalateResponses } from "./escalateResponses";

export class EscalateDialog extends EnterpriseDialog {
    
    // Fields
    public static readonly _responder: EscalateResponses = new EscalateResponses();

    constructor(botServices: BotServices) {
        super(botServices, EscalateDialog.name);
        this.initialDialogId = EscalateDialog.name;

        const escalate = [
            EscalateDialog.SendPhone.bind(this),
        ];

        this.addDialog(new WaterfallDialog(this.initialDialogId, escalate));
    }

    private static async SendPhone(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        await EscalateDialog._responder.replyWith(sc.context, EscalateResponses.ResponseIds.SendPhoneMessage);
        return await sc.endDialog();
    }
}

File: templates/Virtual-Assistant-Template/VirtualAssistantTemplate/VirtualAssistantTemplate/Dialogs/Escalate/EscalateDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using VirtualAssistantTemplate.Dialogs.Shared;
using Microsoft.Bot.Builder.Dialogs;

namespace VirtualAssistantTemplate.Dialogs.Escalate
{
    public class EscalateDialog : EnterpriseDialog
    {
        private EscalateResponses _responder = new EscalateResponses();

        public EscalateDialog(BotServices botServices)
            : base(botServices, nameof(EscalateDialog))
        {
            InitialDialogId = nameof(EscalateDialog);

            var escalate = new WaterfallStep[]
            {
                SendPhone,
            };

            AddDialog(new WaterfallDialog(InitialDialogId, escalate));
        }

        private async Task<DialogTurnResult> SendPhone(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            await _responder.ReplyWith(sc.Context, EscalateResponses.ResponseIds.SendPhoneMessage);
            return await sc.EndDialogAsync();
        }
    }
}
